### PR TITLE
using Classpath shrinker to warn on indirect targets

### PR DESCRIPTION
**Pull request by nadavwe**
this is written in top of https://github.com/bazelbuild/rules_scala/pull/227.
@natansil - we should probably rebase on master.
this still fails - we have something unexpected regarding the output targets, need to investigate.

@ittaiz @or-shachar

### CODE REVIEW

**Comment by nadavwe** [comment]
relates to #235

**Comment by natansil** [comment]
@nadavwe , there were two bugs related to string encoding/escaping.
first, we encoded the jars and targets with ":" instead of "," to the temp file read by the worker.
second, specifically for targets, they contain ":", themselves, so they needed to be escaped...

both are fixed now

**Comment by ittaiz** [comment]
1) would you like me to review?
2) are you copying the code in? If not should I review it in the other repo?

**Comment by nadavwe** [comment]
@ittaiz 
1. please do - i think this is a bit too POC code yet - it should have major beautification refactoring.
2. please review in the other repository anyway and we can decide on copying later.

**Comment by ittaiz** [comment]
Will do

File: plugin/src/main/resources/toolbox.classpath [hunk 1]
```diff
@@ -0,0 +1 @@
+/Users/natans/Work/rules_scala/bazel-genfiles/external/scala/_ijar/scala-compiler/external/scala/lib/scala-compiler-ijar.jar:/Users/natans/Work/rules_scala/bazel-genfiles/external/scala/_ijar/scala-library/external/scala/lib/scala-library-ijar.jar
```
Comment (ittaiz) [inline plugin/src/main/resources/toolbox.classpath]: why do you need this? only for tests? take a look at $location expansion on bazel.build

File: plugin/src/test/scala/io/github/retronym/dependencyanalyzer/DependencyAnalyzerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  import Dependencies._
+
+  @Test
+  def `error on indirect dependency target`(): Unit = {
+    val testCode =
+      """object Foo {
+        |  org.apache.commons.lang3.ArrayUtils.EMPTY_BOOLEAN_ARRAY.length
+        |}
+      """.stripMargin
+    val commonsPath = Coursier.getArtifact(commons)
```
Comment (ittaiz) [inline plugin/src/test/scala/io/github/retronym/dependencyanalyzer/DependencyAnalyzerSpec.scala]: this looks like a really problematic pattern for a bazel build. why not have "commons" in the deps?

File: plugin/src/test/BUILD [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+scala_junit_test(
+        name = "dependency_analyzer_test",
+        srcs = ["scala/io/github/retronym/dependencyanalyzer/DependencyAnalyzerSpec.scala",
+                "scala/io/github/retronym/dependencyanalyzer/TestUtil.scala"],
+        suffixes = ["Test"],
+        size = "small",
+        deps = ["//main:dependency_analyzer",
+                "@io_get_coursier_coursier_cache//jar",
+                "@io_get_coursier_coursier//jar",
+                "@org_scalaz_scalaz_concurrent_2_11//jar",
```
Comment (ittaiz) [inline plugin/src/test/BUILD]: why do you need this? also, why do you need the scalaz stuff?

**Comment by nadavwe** [comment]
changes were merged to upstream in https://github.com/bazelbuild/rules_scala/pull/243.

**Pull request closed without merging**